File: apps/desktop/src-tauri/src/application/rbac_service.rs
use std::sync::Arc;

use crate::domain::entities::{GuildUserRole, MyRole};
use crate::domain::ports::RbacRepository;

pub struct RbacService {
    repo: Arc<dyn RbacRepository>,
}

impl RbacService {
    pub fn new(repo: Arc<dyn RbacRepository>) -> Self {
        Self { repo }
    }

    pub async fn list_guild_users(&self, guild_id: String) -> Result<Vec<GuildUserRole>, String> {
        self.repo.list_guild_users(guild_id).await
    }

    pub async fn get_my_role(&self, guild_id: String) -> Result<MyRole, String> {
        self.repo.get_my_role(guild_id).await
    }

    pub async fn grant_role(
        &self,
        guild_id: String,
        user_id: String,
        role: String,
        display_name: Option<String>,
    ) -> Result<(), String> {
        self.repo.grant_role(guild_id, user_id, role, display_name).await
    }

    pub async fn update_role(
        &self,
        guild_id: String,
        user_id: String,
        role: String,
    ) -> Result<(), String> {
        self.repo.update_role(guild_id, user_id, role).await
    }

    pub async fn revoke_role(&self, guild_id: String, user_id: String) -> Result<(), String> {
        self.repo.revoke_role(guild_id, user_id).await
    }
}

File: apps/web/src/services/rbacService.ts
import { httpGet, httpPost, httpPatch, httpDelete, httpPut } from "@/api/http";
import type { ComponentVisibilityEntry, GuildUserRole, MyRole, RbacRole } from "@/types";

export const rbacService = {
  listGuildUsers(guildId: string): Promise<GuildUserRole[]> {
    return httpGet(`/api/rbac/guilds/${guildId}/users`);
  },
  getMyRole(guildId: string): Promise<MyRole> {
    return httpGet(`/api/rbac/me/${guildId}`);
  },
  grantRole(guildId: string, userId: string, role: RbacRole, displayName?: string | null): Promise<unknown> {
    return httpPost(`/api/rbac/guilds/${guildId}/users/${userId}`, { role, display_name: displayName ?? null });
  },
  updateRole(guildId: string, userId: string, role: RbacRole): Promise<unknown> {
    return httpPatch(`/api/rbac/guilds/${guildId}/users/${userId}`, { role });
  },
  revokeRole(guildId: string, userId: string): Promise<void> {
    return httpDelete(`/api/rbac/guilds/${guildId}/users/${userId}`);
  },
  listComponentVisibility(guildId: string): Promise<ComponentVisibilityEntry[]> {
    return httpGet(`/api/rbac/component-visibility/${guildId}`);
  },
  upsertComponentVisibility(guildId: string, entries: ComponentVisibilityEntry[]): Promise<unknown> {
    return httpPut(`/api/rbac/component-visibility/${guildId}`, { entries });
  },
};
